packages/core/src/input/input-file-resolver.ts

```diff
@@ -119,58 +74,92 @@ export class InputFileResolver {
+        if (logAboutUselessPatterns && files.length === 0) {
+          this.log.warn(`Glob pattern "${pattern}" did not exclude any files.`);
+        }
         files.forEach((fileName) => fileSet.delete(fileName));
       } else {
-        const files = await this.expandPattern(pattern, logAboutUselessPatterns);
+        const files = this.filterPattern(fileNames, pattern);
+        if (logAboutUselessPatterns && files.length === 0) {
+          this.log.warn(`Glob pattern "${pattern}" did not result in any files.`);
+        }
         files.forEach((fileName) => fileSet.add(fileName));
       }
     }
     return Array.from(fileSet);
   }
 
-  private async expandPattern(globbingExpression: string, logAboutUselessPatterns: boolean): Promise<string[]> {
-    if (MUTATION_RANGE_REGEX.exec(globbingExpression)) {
-      globbingExpression = globbingExpression.replace(MUTATION_RANGE_REGEX, '$1');
+  private filterPattern(fileNames: Iterable<string>, pattern: string): string[] {
+    if (MUTATION_RANGE_REGEX.exec(pattern)) {
+      pattern = pattern.replace(MUTATION_RANGE_REGEX, '$1');
     }
+    const matcher = new FileMatcher(pattern);
 
-    const fileNames = (await glob(globbingExpression)).map((relativeFile) => path.resolve(relativeFile));
-    if (!fileNames.length && logAboutUselessPatterns) {
-      this.log.warn(`Globbing expression "${globbingExpression}" did not result in any files.`);
-    }
-    return fileNames;
+    const filteredFileNames = [...fileNames].filter((fileName) => matcher.matches(fileName));
+    return filteredFileNames;
   }
 
-  private async resolveFilesUsingGit(): Promise<string[]> {
-    try {
-      const { stdout } = await childProcessAsPromised.exec(`git ls-files --others --exclude-standard --cached --exclude /${this.tempDirName}/*`, {
-        maxBuffer: 10 * 1000 * 1024,
-      });
-      const relativeFileNames = stdout
-        .toString()
-        .split('\n')
-        .map((line) => line.trim())
-        .filter(Boolean) // remove empty lines
-        .map(decodeGitLsOutput);
-
-      const fileNames = relativeFileNames.map((relativeFileName) => path.resolve(relativeFileName));
-      return fileNames;
-    } catch (error) {
-      throw new StrykerError(
-        normalizeWhitespaces(
-          `Cannot determine input files. Either specify a \`files\`
-        array in your stryker configuration, or make sure "${process.cwd()}"
-        is located inside a git repository`
-        ),
-        error
+  private async resolveInputFiles(): Promise<string[]> {
+    const ignoreRules = this.ignoreRules.map((pattern) => new Minimatch(pattern, { dot: true, flipNegate: true, nocase: true }));
+
+    // Inspired by https://github.com/npm/ignore-walk/blob/0e4f87adccb3e16f526d2e960ed04bdc77fd6cca/index.js#L124
+    const matchesDirectory = (entryName: string, entryPath: string, rule: IMinimatch) => {
+      return (
+        matchesFile(entryName, entryPath, rule) ||
```

Comment:
This is quite unreadable as you're combining a lot of matches. It might be more readable if the sets are more on one line (for example `rule.negate && (rule.match(`/${entryPath}`, true) || rule.match(entryPath, true))`

It might also be smart to place the checks in separate variables so you can actually name what's going on

Reply:
Since we're using `prettier` I can't play around with the formatting. Separate variables would work, but it would technically cost performance since all matches would always be done (instead of stopping at the first failing match).

I've added a third `matchesXxx` helper function:

```js
const matchesDirectoryPartially = (entryPath: string, rule: IMinimatch) => {
      // @ts-expect-error Missing overload in type definitions. See https://github.com/isaacs/minimatch/issues/134
      return rule.match(`/${entryPath}`, true) || rule.match(entryPath, true);
    };
```

This helps a lot I think.